### PR TITLE
Add CVE-2024-40725

### PR DESCRIPTION
**Pull request by nmasdoufi-ol**

### CODE REVIEW

File: agent/exploits/cve_2024_40725.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
+"""Agent implementation for detecting source code disclosure vulnerability in Apache HTTP Server."""
+
+import requests
+import re
+import os
+import logging
+from semantic_version import Version
```
Comment (3asm) [inline agent/exploits/cve_2024_40725.py:7]: No direct class import. Applies to all the ones below.

File: agent/exploits/cve_2024_40725.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+    "A source code disclosure vulnerability in Apache HTTP Server 2.4.60 and 2.4.61 allows for the exposure of local "
+    "content under certain circumstances. This issue affects the legacy content-type based configuration of handlers."
+)
+DEFAULT_TIMEOUT = 30
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+def get_server_version(url: str) -> Optional[str]:
```
Comment (3asm) [inline agent/exploits/cve_2024_40725.py:32]: Private method.

File: agent/exploits/cve_2024_40725.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+        True if the version is vulnerable, else False.
+    """
+    if version is None:
+        return False
+    try:
+        version_obj = Version(version)
+        vulnerable_start = Version("2.4.60")
+        vulnerable_end = Version("2.4.61")
+        return bool(vulnerable_start <= version_obj <= vulnerable_end)
+    except Exception as e:
```
Comment (3asm) [inline agent/exploits/cve_2024_40725.py:70]: No generic exception catching.

File: agent/exploits/cve_2024_40725.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+        logger.error("Error parsing version: %s", e)
+        return False
+
+
+def read_filenames(file_path: str) -> List[str]:
```
Comment (3asm) [inline agent/exploits/cve_2024_40725.py:75]: Private method.

File: agent/exploits/cve_2024_40725.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+
+    try:
+        with open(absolute_path, "r") as file:
+            return [line.strip() for line in file if line.strip()]
+    except IOError as e:
+        logger.error("Error reading file list: %s", e)
+        return []
+
+
+def highlight_code(content: str) -> bool:
```
Comment (3asm) [inline agent/exploits/cve_2024_40725.py:96]: Wrong method method.

File: agent/exploits/cve_2024_40725.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+                "No filenames to check. Please ensure filenames.txt exists and is correctly formatted."
+            )
+            return []
+
+        vulnerabilities = []
+        if check_files(target.origin, filenames):
+            vulnerability = _create_vulnerability(target)
+            vulnerabilities.append(vulnerability)
+        else:
+            logger.info("No source code disclosure detected with the tested files.")
```
Comment (3asm) [inline agent/exploits/cve_2024_40725.py:221]: Many of those lines should have a test case.

**Review by 3asm — CHANGES_REQUESTED** [review]

File: .mypy.ini [hunk 7]
```diff
@@ -20,6 +20,7 @@ warn_unused_ignores = True
 disallow_any_unimported = True
 warn_return_any = True
 exclude = .*_pb2.py
+ignore_missing_imports = True
```
Comment (PiranhaSa) [inline .mypy.ini:23]: why is this ?

Comment (nmasdoufi-ol) [reply .mypy.ini:23]: error: Skipping analyzing "semantic_version": module is installed, but missing library stubs or py.typed marker  [import-untyped]

https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports

File: agent/exploits/cve_2024_40725.py (hunk 6 above)
Comment (nmasdoufi-ol) [reply agent/exploits/cve_2024_40725.py:227]: Done.

File: agent/exploits/cve_2024_40725.py [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+    Returns:
+        The server version string if found, else None.
+    """
+    try:
+        response = requests.head(url, verify=False)
+        server_header = response.headers.get("Server", "")
+        if "Apache" in server_header:
+            version_match = re.search(r"Apache/(\d+\.\d+\.\d+)", server_header)
+            if version_match:
+                return version_match.group(1)
```
Comment (elyousfi5) [inline agent/exploits/cve_2024_40725.py:48]:
```suggestion
            if version_match is True:
                return version_match.group(1)
```

File: agent/exploits/cve_2024_40725.py [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+
+    Args:
+        content: The content to check.
+
+    Returns:
+        True if the content contains code, else False.
+    """
+    try:
+        lexer = guess_lexer(content)
+        if not isinstance(lexer, (HtmlLexer, CssLexer)):
```
Comment (PiranhaSa) [inline agent/exploits/cve_2024_40725.py:108]:
```suggestion
        if  isinstance(lexer, (HtmlLexer, CssLexer)) is False:
```

File: agent/exploits/cve_2024_40725.py [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+    for filename in filenames:
+        file_url = url + filename
+        try:
+            response = requests.get(file_url, verify=False)
+            response.raise_for_status()
+            if any(file_url.endswith(ext) for ext in EXTENSIONS):
+                logger.info("Potential source code disclosure detected!")
+                logger.info("File: %s", filename)
+                logger.info("URL: %s\n", file_url)
+                if _contains_code_content(response.text):
```
Comment (PiranhaSa) [inline agent/exploits/cve_2024_40725.py:137]:
```suggestion
                if _contains_code_content(response.text) is True:
```

File: agent/exploits/cve_2024_40725.py (hunk 8 above)
Comment (nmasdoufi-ol) [reply agent/exploits/cve_2024_40725.py:48]: error: Non-overlapping identity check (left operand type: "str | None", right operand type: "Literal[True]")  [comparison-overlap]

File: agent/exploits/cve_2024_40725.py [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+
+import requests
+import re
+from os import getcwd
+from os.path import abspath
+import logging
+from semantic_version import Version
+from pygments import highlight
+from pygments.lexers import guess_lexer, HtmlLexer, CssLexer
+from pygments.formatters import TerminalFormatter
```
Comment (3asm) [inline agent/exploits/cve_2024_40725.py:11]: No direct classes import.

File: agent/exploits/cve_2024_40725.py [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+from typing import Optional, List
```
Comment (3asm) [inline agent/exploits/cve_2024_40725.py:12]: Use | None and list.

File: agent/exploits/cve_2024_40725.py [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+
+    Returns:
+        True if the version is vulnerable, else False.
+    """
+    if version is None:
+        return False
+    try:
+        version_obj = Version(version)
+        vulnerable_start = Version("2.4.60")
+        vulnerable_end = Version("2.4.61")
```
Comment (3asm) [inline agent/exploits/cve_2024_40725.py:69]: Const.

**Review by 3asm — CHANGES_REQUESTED** [review]

File: agent/exploits/cve_2024_40725.py [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+from pygments import highlight
+from pygments.lexers import guess_lexer, HtmlLexer, CssLexer
+from pygments.formatters import TerminalFormatter
+from typing import Optional, List
+
+from agent import definitions
+from agent import exploits_registry
+
+from ostorlab.agent.kb import kb
+from ostorlab.agent.mixins import agent_report_vulnerability_mixin
```
Comment (rabsondev) [inline agent/exploits/cve_2024_40725.py:18]:
```suggestion
from ostorlab.agent.kb import kb
from ostorlab.agent.mixins import agent_report_vulnerability_mixin

from agent import definitions
from agent import exploits_registry
```

File: agent/exploits/cve_2024_40725.py (hunk 13 above)
Comment (nmasdoufi-ol) [reply agent/exploits/cve_2024_40725.py]: Done.

File: .mypy.ini (hunk 7 above)
Comment (deadly-panda) [inline .mypy.ini:23]: we usually apply it for that specific package.. like `scipy`, `adb_shell` below ..

File: agent/exploits/cve_2024_40725.py [hunk 15]
```diff
@@ -0,0 +1,229 @@
+"""Agent implementation for detecting source code disclosure vulnerability in Apache HTTP Server."""
+
+import os
+import requests
+import re
+import logging
+import semantic_version
+from pygments import highlight
```
Comment (deadly-panda) [inline agent/exploits/cve_2024_40725.py:8]: this should be added to `requirements.txt`

File: agent/exploits/cve_2024_40725.py [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+        content: The content to check.
+
+    Returns:
+        True if the content contains code, else False.
+    """
+    try:
+        lexer = lexers.guess_lexer(content)
+        if not isinstance(lexer, (lexers.HtmlLexer, lexers.CssLexer)):
+            highlighted = highlight(content, lexer, formatters.TerminalFormatter())
+            logger.info(highlighted)
```
Comment (deadly-panda) [inline agent/exploits/cve_2024_40725.py:110]:
```suggestion
            logger.info("%s", highlighted)
```

File: agent/exploits/cve_2024_40725.py [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+            response = requests.get(file_url, verify=False)
+            response.raise_for_status()
+            if any(file_url.endswith(ext) for ext in EXTENSIONS):
+                logger.info("Potential source code disclosure detected!")
+                logger.info("File: %s", filename)
+                logger.info("URL: %s\n", file_url)
+                if _contains_code_content(response.text):
+                    return True
+                else:
+                    logger.info("Content is not source code.")
```
Comment (deadly-panda) [inline agent/exploits/cve_2024_40725.py:140]: I think this may be a bit too noisy..

File: agent/exploits/cve_2024_40725.py [hunk 18]
```diff
@@ -0,0 +1,229 @@
...
+        """Checks if the target is running a vulnerable version of Apache HTTP Server.
+
+        Args:
+            target: The target to check.
+
+        Returns:
+            True if the target is running a vulnerable version, else False.
+        """
+        version = _get_server_version(target.origin)
+        if version:
```
Comment (deadly-panda) [inline agent/exploits/cve_2024_40725.py:202]:
```suggestion
        if version is not None:
```

File: .mypy.ini (hunk 7 above)
Comment (nmasdoufi-ol) [reply .mypy.ini:23]: I was getting this error from mypy

`error: Skipping analyzing "semantic_version": module is installed, but missing library stubs or py.typed marker [import-untyped]`

File: agent/exploits/cve_2024_33544.py [hunk 19]
```diff
@@ -0,0 +1,32 @@
+"""Agent Asteroid implementation for CVE-2024-29269"""
+
+import re
+from agent import definitions
```
Comment (deadly-panda) [inline agent/exploits/cve_2024_33544.py:4]:
```suggestion
import re

from agent import definitions
```

File: agent/exploits/cve_2024_40725.py [hunk 20]
```diff
@@ -0,0 +1,227 @@
+"""Agent implementation for detecting source code disclosure vulnerability in Apache HTTP Server."""
+
+import os
+import requests
+import re
+import logging
+import semantic_version
+from pygments import highlight
+from pygments import lexers
+from pygments import formatters
```
Comment (deadly-panda) [inline agent/exploits/cve_2024_40725.py:10]:
```suggestion
import os
import re
import logging

import requests
import semantic_version
from pygments import highlight
from pygments import lexers
from pygments import formatters
```

File: agent/exploits/cve_2024_40725.py [hunk 21]
```diff
@@ -0,0 +1,227 @@
...
+
+from ostorlab.agent.kb import kb
+from ostorlab.agent.mixins import agent_report_vulnerability_mixin
```
Comment (deadly-panda) [inline agent/exploits/cve_2024_40725.py:13]:
```suggestion
import os
import re
import logging

import semantic_version
import requests
from pygments import highlight
from pygments import lexers
from pygments import formatters
from ostorlab.agent.kb import kb
from ostorlab.agent.mixins import agent_report_vulnerability_mixin
```

File: agent/exploits/cve_2024_40725.py [hunk 22]
```diff
@@ -0,0 +1,227 @@
...
+    Returns:
+        True if the version is vulnerable, else False.
+    """
+    if version is None:
+        return False
+    try:
+        version_obj = semantic_version.Version(version)
+        return bool(VULNERABLE_START <= version_obj <= VULNERABLE_END)
+    except ValueError as e:
+        logger.error("Error parsing version: %s", e)
```
Comment (deadly-panda) [inline agent/exploits/cve_2024_40725.py:72]:
```suggestion
        logger.error("Error parsing version: %s : %s", version, e)
```

File: agent/exploits/cve_2024_40725.py [hunk 23]
```diff
@@ -0,0 +1,227 @@
...
+    def check(self, target: definitions.Target) -> list[definitions.Vulnerability]:
+        """Checks for source code disclosure vulnerability in Apache HTTP Server.
+
+        Args:
+            target: The target to check.
+
+        Returns:
+            A list of vulnerabilities found.
+        """
+        filenames = _read_filenames(os.path.abspath("Wordlists/filenames.txt"))
```
Comment (deadly-panda) [inline agent/exploits/cve_2024_40725.py:214]: Please use pathlib..

File: agent/exploits/cve_2024_40725.py [hunk 24]
```diff
@@ -0,0 +1,227 @@
...
+        if not filenames:
```
Comment (deadly-panda) [inline agent/exploits/cve_2024_40725.py:215]:
```suggestion
        if len(filenames) == 0:
```

File: agent/exploits/cve_2024_40725.py (hunk 10 above)
Comment (PiranhaSa) [reply agent/exploits/cve_2024_40725.py:137]: ?

File: .mypy.ini (hunk 7 above)
Comment (ErebusZ) [reply .mypy.ini:28]: he means you apply it on the specific package thats triggering the mypy error, check other agents for reference.

**Review by adnaneserrar — APPROVED** [review]

**Review by 3asm — APPROVED** [review]

**Pull request merged**